fastpass/src/main/scala/scala/meta/internal/fastpass/pantsbuild/CompileBFS.scala

```diff
@@ -10,77 +10,121 @@ import scala.meta.internal.fastpass.pantsbuild.commands.StrictDepsMode
+
         val newDepth =
-          if (dep.pantsTargetType.isTarget) depth
+          if (target.pantsTargetType.isTarget) depth
           else depth - 1
-        dep.dependencies.foreach(d => loop(export.targets(d), newDepth))
+
+        val depsDepthMaps = deps.map(depthMap(_, newDepth))
+        val exportsDepthMaps = deps.map(exportDepthMap)
+
+        val maxDepth = Math.max(
+          depsDepthMaps.map(_.size).max,
+          exportsDepthMaps.map(_.size).max
+        )
+
+        (0 until maxDepth).foreach { currentDepth =>
+          val level = new mutable.ArrayBuffer[PantsTarget]()
+
+          // unlike exports, dependencies are followed only until `depth`.
+          if (currentDepth < depth) {
+            depsDepthMaps.foreach { depthMap =>
+              if (depthMap.size > currentDepth) {
+                level ++= depthMap(currentDepth).filter(seen.add)
+              }
+            }
+          }
+
+          exportsDepthMaps.foreach { depthMap =>
+            if (depthMap.size > currentDepth) {
+              level ++= depthMap(currentDepth).filter(seen.add)
+            }
+          }
+
+          if (level.nonEmpty) result += level
+        }
       }
-      loop(target, mode.plusDepth)
       result
     }
   }
 
-  private def dependencyExports(name: String): Iterable[PantsTarget] = {
-    if (isInProgress.contains(name)) {
-      throw new IllegalArgumentException(s"illegal cycle at target '$name'")
-    }
-
-    def uncached(): Iterable[PantsTarget] = {
-      val dep = export.targets(name)
-      if (!dep.scope.isCompile) Nil
-      else {
-        isInProgress.add(name)
-        val result = new mutable.LinkedHashSet[PantsTarget]()
-        val exportNames: Seq[String] =
-          if (
-            dep.pantsTargetType.isTarget || dep.pantsTargetType.isJarLibrary
-          ) {
-            // NOTE(olafur): it seems like `target` and `jar_library` types
-            // export their dependencies even if they have an empty `exports`
-            // field. This should probably be reflected in the output of
-            // `export-fastpass` but for now it's OK to do this logic on the
-            // fastpass-side instead.
-            dep.dependencies
-          } else {
-            for {
-              dependencyName <- dep.dependencies
-              if dep.exports.contains(dependencyName) ||
-                // NOTE(olafur): all synthetic dependencies are automatically
-                // exported if they are derived from this target. We don't check
-                // that this synthetic target is derived from this target, which
-                // might cause subtle bugs that we may need to fix in the
-                // future.
-                export.targets(dependencyName).isSynthetic
-            } yield dependencyName
+  private def exportDepthMap(target: PantsTarget): DepthMap = {
+    def uncached(): DepthMap = {
+      val result = new mutable.ArrayBuffer[mutable.ArrayBuffer[PantsTarget]]
```

Comment:
I think this also works
```suggestion
      val result = new DepthMap
```